### ISSUE TITLE
Failed to upgrade to Kotlin 1.9

### ISSUE THREAD

**Issue by hantsy** [google/ksp#1445]
I updated my project to Kotlin 1.9.0 and Gradle 8.2, and got the following build info.

```bash 
A problem was found with the configuration of task ':kaptGenerateStubsKotlin' (type 'KaptGenerateStubsTask').
  - Gradle detected a problem with the following location: 'D:\FreelanceProjects\sami\etip-backend\build\generated\ksp\main\kotlin'.

    Reason: Task ':kaptGenerateStubsKotlin' uses this output of task ':kspKotlin' without declaring an explicit or implicit dependency. This can lead to incorrect results being produce
d, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':kspKotlin' as an input of ':kaptGenerateStubsKotlin'.
      2. Declare an explicit dependency on ':kspKotlin' from ':kaptGenerateStubsKotlin' using Task#dependsOn.

```

Add the following in the `build.gradle.kts`, did not resolve the issue.

```kotlin 

tasks.named("kaptGenerateStubsKotlin"){
    dependsOn("kspKotlin")
}
```

**Comment by ting-yuan** [comment]
Would you mind sharing a small reproducible test case?

**Comment by hantsy** [comment]
Add this to resolve our problem, our project used Kotlin Dataframe[1] which used Ksp.

```kotlin
tasks.whenTaskAdded {
    if (name == "kaptGenerateStubsKotlin") {
        dependsOn("kspKotlin")
    }
    if (name == "kaptTestKotlin") {
        dependsOn("kspTestKotlin")
    }

//...
}
```
[1]:[https://github.com/Kotlin/dataframe](https://github.com/Kotlin/dataframe)

**Comment by Enaium** [comment]
+1 same here

**Comment by uladzislau-hryshechka** [comment]
+1 same here. We have multimodule android project. We use kapt and ksp at same time. Maybe you could provide a temporary workaround to fix this issue? Thanks in advance! @ting-yuan

**Comment by neetopia** [comment]
@Enaium @uladzislau-hryshechka can you provide mode details on how your project is configured? From @hantsy 's case, it does seem like applying dataframe added a dependency since dataframe generates code with KSP that might be taken as input to KAPT.

**Comment by brentwatson** [comment]
+1 experiencing the same in our project. Same workaround resolved this for us.

**Comment by Enaium** [comment]
> @Enaium @uladzislau-hryshechka can you provide mode details on how your project is configured? From @hantsy 's case, it does seem like applying dataframe added a dependency since dataframe generates code with KSP that might be taken as input to KAPT.

https://github.com/babyfish-ct/jimmer/tree/main/example/kotlin/jimmer-sql-kt

```
kotlin("jvm") version "1.9.0"
kotlin("kapt") version "1.9.0"
kotlin("plugin.spring") version "1.9.0"
id("com.google.devtools.ksp") version "1.9.0-1.0.12"
```

**Comment by uladzislau-hryshechka** [comment]
> @Enaium @uladzislau-hryshechka can you provide mode details on how your project is configured? From @hantsy 's case, it does seem like applying dataframe added a dependency since dataframe generates code with KSP that might be taken as input to KAPT.

We have multimodule android project. Structure of this project:
```
- kotlin library module
- android library module
- app module  (which include all android and kotlin libraries)
```

Some modules could include ksp, kapt, both or nothing of that. It depends. Classical type of plugins which we use (for example android library):

```
plugins {
  id("base.android") - our convention plugin 
  id("kotlin-kapt")
  id("com.google.devtools.ksp")
}
```

**Comment by ZacSweers** [comment]
I don't think this is a bug per-se, rather a side effect of kapt tasks no longer being eagerly initialized. I think the right solution would be for KSP to offer an API to add this dependency at the Gradle level.

Perhaps something like this? Or a gradle property?

```kotlin
ksp {
  makeKaptTasksDependOnKspTasks()
}
```

Where things get tricky are matching variants and KMP jvm source sets, but I think it'd be worthwhile to do.

**Comment by Enaium** [comment]
> I guess some of the issues come from a previous workaround which lets IDE be aware of generated source:
> 
> ```
> kotlin {
> 	sourceSets.main {
> 		kotlin.srcDir("build/generated/ksp/main/kotlin")
> 	}
> }
> ```
> 
> KSP has been doing this automatically, with dependencies correctly set up since 1.0.9. If you have this workaround in your build script, please try to remove it and see if the problem remains.

But it still has problems after run the `clean` task

**Comment by ting-yuan** [comment]
Do you mean that, after removing that block / workaround and doing a clean build, the problem remains? If so, could you provide a sample project so that we can investigate?

**Comment by dimsuz** [comment]
I don't have the `srcDirs` workaround mentioned above, but still got this error after upgrading to kotlin 1.9.10 and the corresponding ksp version.

**Comment by Enaium** [comment]
I solved the issue. 

The kapt plugin must be declared after the ksp plugin.

```diff
plugins {
kotlin("jvm") version "1.9.0"
-	kotlin("kapt") version "1.9.0"
id("com.google.devtools.ksp") version "1.9.0-1.0.12"
+	kotlin("kapt") version "1.9.0"
}
```

**Comment by neetopia** [comment]
> I got this warning in upgrading to kotlin 1.9.10
> 
> ksp-1.8.10-1.0.9 is too old for kotlin-1.9.10. Please upgrade ksp or downgrade kotlin-gradle-plugin to 1.8.10.
> 
> is there any ksp version that is compatible with the kotlin 1.9.10

you can try KSP 1.9.10-1.0.13

**Comment by shivam-gupta007** [comment]
Thanks @neetopia 🙌🙌

**Comment by Lysander** [comment]
Maybe my comment could help you: https://github.com/google/ksp/issues/963#issuecomment-1894144639
